html_orig/jssrc/term_screen.js

```diff
@@ -1,378 +1,645 @@
+const SEQ_SET_COLOR_ATTR = 1
+const SEQ_REPEAT = 2
+const SEQ_SET_COLOR = 3
+const SEQ_SET_ATTR = 4
+
+const SELECTION_BG = '#b2d7fe'
+const SELECTION_FG = '#333'
+
+const themes = [
+  [ // Tango
+    '#111213', '#CC0000', '#4E9A06', '#C4A000', '#3465A4', '#75507B', '#06989A',
+    '#D3D7CF',
+    '#555753', '#EF2929', '#8AE234', '#FCE94F', '#729FCF', '#AD7FA8', '#34E2E2',
+    '#EEEEEC'
+  ],
+  [ // Linux
+    '#000000', '#aa0000', '#00aa00', '#aa5500', '#0000aa', '#aa00aa', '#00aaaa',
+    '#aaaaaa',
+    '#555555', '#ff5555', '#55ff55', '#ffff55', '#5555ff', '#ff55ff', '#55ffff',
+    '#ffffff'
+  ],
+  [ // xterm
+    '#000000', '#cd0000', '#00cd00', '#cdcd00', '#0000ee', '#cd00cd', '#00cdcd',
+    '#e5e5e5',
+    '#7f7f7f', '#ff0000', '#00ff00', '#ffff00', '#5c5cff', '#ff00ff', '#00ffff',
+    '#ffffff'
+  ],
+  [ // rxvt
+    '#000000', '#cd0000', '#00cd00', '#cdcd00', '#0000cd', '#cd00cd', '#00cdcd',
+    '#faebd7',
+    '#404040', '#ff0000', '#00ff00', '#ffff00', '#0000ff', '#ff00ff', '#00ffff',
+    '#ffffff'
+  ],
+  [ // Ambience
+    '#2e3436', '#cc0000', '#4e9a06', '#c4a000', '#3465a4', '#75507b', '#06989a',
+    '#d3d7cf',
+    '#555753', '#ef2929', '#8ae234', '#fce94f', '#729fcf', '#ad7fa8', '#34e2e2',
+    '#eeeeec'
+  ],
+  [ // Solarized
+    '#073642', '#dc322f', '#859900', '#b58900', '#268bd2', '#d33682', '#2aa198',
+    '#eee8d5',
+    '#002b36', '#cb4b16', '#586e75', '#657b83', '#839496', '#6c71c4', '#93a1a1',
+    '#fdf6e3'
+  ]
+]
+
+class TermScreen {
+  constructor () {
+    this.canvas = document.createElement('canvas')
+    this.ctx = this.canvas.getContext('2d')
+
+    if ('AudioContext' in window || 'webkitAudioContext' in window) {
+      this.audioCtx = new (window.AudioContext || window.webkitAudioContext)()
+    } else {
+      console.warn('No AudioContext!')
+    }
+
+    this.cursor = {
+      x: 0,
+      y: 0,
+      fg: 7,
+      bg: 0,
+      attrs: 0,
+      blinkOn: false,
+      visible: true,
+      hanging: false,
+      style: 'block',
+      blinkInterval: null
+    }
+
+    this._colors = themes[0]
+
+    this._window = {
+      width: 0,
+      height: 0,
+      devicePixelRatio: 1,
+      fontFamily: '"DejaVu Sans Mono", "Liberation Mono", "Inconsolata", ' +
+        'monospace',
+      fontSize: 20,
+      gridScaleX: 1.0,
+      gridScaleY: 1.2,
+      blinkStyleOn: true,
+      blinkInterval: null
+    }
+    this.windowState = {
+      width: 0,
+      height: 0,
+      devicePixelRatio: 0,
+      gridScaleX: 0,
+      gridScaleY: 0,
+      fontFamily: '',
+      fontSize: 0
+    }
+    this.selection = {
+      selectable: true,
+      start: [0, 0],
+      end: [0, 0]
+    }
+
+    this.mouseMode = { clicks: false, movement: false }
+
+    const self = this
+    this.window = new Proxy(this._window, {
+      set (target, key, value, receiver) {
+        target[key] = value
+        self.updateSize()
+        self.scheduleDraw()
+        return true
+      }
+    })
```

Comment:
I'm not sure what this is for, but it doesn't work with Android Browser (non-chrome). Is it necessary?

Reply:
That proxy ensures that when you write to the window object the size is updated and the terminal is redrawn. According to caniuse the Android Browser doesn't support any part of this file so the proxy isn't really the issue, so I'll either have to rewrite this to use ES5 or run this through Babel. For what it's worth, the Babel version is only half a kilobyte larger.